### ISSUE TITLE
RaspberryPi status

### ISSUE THREAD

**Issue by Abdur-rahmaanJ** [joshiemoore/snakeware#10]
Greetings,

Can it be run on the Pi?

**Comment by ankith26** [comment]
You would have to built it from source on a pi. But if you just want to test, run wm.py in the snakewm directory.

**Comment by Abdur-rahmaanJ** [comment]
Ah ok. Else i just PRed a clock the right way

**Comment by ADVFRED** [comment]
How would you go about building snake ware on a pi?(Im a general newbie to Linux/Unix)

**Comment by Admicos** [comment]
On Sat May 30, 2020 at 6:03 PM +03, ADVFRED wrote:
> How would you go about building snake ware on a pi?(Im a general newbie
> to Linux/Unix)

With the current build system, I'd imagine you'd need to build it ON a
Pi, as the current build scripts copy files from the host OS. And you
would probably also need to modify the kernel configuration, as it
probably assumes an x86 (common desktop/server) system right now. (The
Raspberry Pi is an 'arm' system)

Even after that, you'd need to mess with boot firmware files as the Pi
doesn't have a built-in boot loader, which can probably be done by
stealing the configuration from another distro like Raspbian or Alpine.

However, if the build system gets switched to something like Buildroot,
which they seem to be looking at, it should be an easier process, as
there are a lot of resources on how to work with Buildroot and a
Raspberry Pi[1], and even if there wasn't, having a "standardized" build
system would mean there could be different "backends" for different
architectures, including the Raspberry Pi, without needing to modify too
much.

[1]: https://www.google.com/search?hl=en&q=buildroot%20raspberry%20pi

**Comment by joshiemoore** [comment]
The original build system has been replaced with a build system based on buildroot, so porting snakeware to Raspberry Pi should be a lot more manageable.

**Comment by alexandruavadanii** [comment]
I can't say I have too much hands-on experience with the Pi, but I have some experience porting various stuff to aarch64.
The first thing that comes to mind when considering porting snakeware (apart from building arch-compatible binaries via buildroot) is the on-target bootloader. Most aarch64 boards (but not the Pi) come with UEFI and only that.
The current snakeware implementation seems to target BIOS platforms (i.e. only x86), so for UEFI compatibility we'd have to change things a bit. Btw, the current snakeware building scripts seem to call the build host's grub binary and expect it to be a BIOS-compatible grub and not an UEFI-compatible grub - that's something that can probably be improved, but I haven't dug too deep.

Anyway, I'd look into "direct kernel booting" as a method to bypass the need to chainload a system loader like grub, and instead execute the kernel binary directly.

As for the kernel/buildroot configuration, I'm sure there are people out there (and not few) who are succesfully using buildroot to create images for the Pi - simply try using one of their configs as a starting point. Alternatively, we could copy the kernel config from another distribution currently supporting the Pi.

**Comment by erkinalp** [comment]
Raspberry Pi is neither BIOS or UEFI. It uses a custom firmware.

**Comment by alexandruavadanii** [comment]
That's great news (Pi supporting UEFI).
If https://github.com/joshiemoore/snakeware/pull/30 gets merged, the changes required for adapting the build should be clearer, although not trivial - use grub-efi-arm64 inside the chroot, add an ESP partition and use GPT for the image (throw in some isohybrid for x86 and get UEFI/BIOS compatible images on x86 for free in the process).

If anybody wants to pick this up and try to implement it, I can probably give some more detailed hints as I recently did something similar in some other OSS repos, e.g. https://github.com/akraino-edge-stack/ta-build-tools/commit/eb6df414a1b5d1b7b67d3dd5f9b7d0529b8cf5fc or https://github.com/akraino-edge-stack/ta-ipa-deployer/commit/d881d27e8eaaf154f24432f4a9c19a10a6fa7701 .

**Comment by erkinalp** [comment]
Anyway, there is an experimental UEFI port for Raspberry Pi 4:
https://github.com/pftf/RPi4

**Comment by alexandruavadanii** [comment]
Hi, @joshiemoore, I see you assigned this to yourself - do you plan on starting working on it right away?
I'm asking because I have been thinking about giving it a try (bought a RPi 4B today, having some free time on Monday). My plan was to base my work on https://github.com/joshiemoore/snakeware/pull/30, start with buildroot's defconfig for RPi4 (64 bit), no grub (the firmware should be able to boot the kernel directly) and cross-compiling on x86.

**Comment by joshiemoore** [comment]
@alexandruavadanii I've already been working on it, I have it running on my pi 3, and have a Pi 4 showing up on Saturday (which is probably what the final Pi target will end up being)

**Comment by joshiemoore** [comment]
@alexandruavadanii but also if you end up experimenting with it too I'd love to hear about your results! Just so you know the default RPI buildroot configs do not work, they use a broken version of the kernel that doesn't compile for some reason. I fixed this by pointing the kernel at the RPI Linux kernel git and setting the version to 4.19.y. After that I built and runs fine. I need to tweak the config more before it's ready for primetime but it's close.
